File: servers/elections/scripts/generate-build-info.exs
#!/usr/bin/env elixir
# Generate BuildInfo.json for Elixir server
#
# Captures:
# - First 8 digits of latest commit hash
# - Branch name
# - Authored date
# - Commit date
# - Build date (with time)

Mix.install([{:jason, "~> 1.2"}])

defmodule BuildInfoGenerator do
  def generate do
    git_info = get_git_info()
    build_date = DateTime.utc_now() |> DateTime.to_iso8601()
    
    build_info = %{
      commitHash: git_info[:commit_hash],
      branch: git_info[:branch],
      authorDate: git_info[:author_date],
      commitDate: git_info[:commit_date],
      buildDate: build_date
    }
    
    # Write to priv/static so it can be served, and also to lib for compile-time access
    json_content = Jason.encode!(build_info, pretty: true)
    
    # Write to priv/static for serving via HTTP
    priv_static_path = Path.join([__DIR__, "..", "priv", "static", "build-info.json"])
    File.mkdir_p!(Path.dirname(priv_static_path))
    File.write!(priv_static_path, json_content)
    
    # Write to lib for compile-time module generation
    lib_path = Path.join([__DIR__, "..", "lib", "elections", "build_info.json"])
    File.write!(lib_path, json_content)
    
    IO.puts("BuildInfo.json generated: #{inspect(build_info)}")
    build_info
  end
  
  defp get_git_info do
    try do
      # Get repo root (3 levels up from scripts: scripts -> elections -> servers -> repo root)
      repo_root = Path.expand(Path.join([__DIR__, "..", "..", ".."]))
      
      # Always run git commands from repo root to ensure consistency with UI script
      # This ensures we get the latest commit hash for the entire repo
      commit_hash = System.cmd("git", ["rev-parse", "HEAD"], cd: repo_root)
                    |> elem(0)
                    |> String.trim()
      short_hash = String.slice(commit_hash, 0, 8)
      
      branch = System.cmd("git", ["rev-parse", "--abbrev-ref", "HEAD"], cd: repo_root)
               |> elem(0)
               |> String.trim()
      
      author_date = System.cmd("git", ["log", "-1", "--format=%ai"], cd: repo_root)
                    |> elem(0)
                    |> String.trim()
      
      commit_date = System.cmd("git", ["log", "-1", "--format=%ci"], cd: repo_root)
                    |> elem(0)
                    |> String.trim()
      
      %{
        commit_hash: short_hash,
        branch: branch,
        author_date: author_date,
        commit_date: commit_date
      }
    rescue
      e ->
        IO.warn("Warning: Could not get git info: #{inspect(e)}")
        now = DateTime.utc_now() |> DateTime.to_iso8601()
        %{
          commit_hash: "unknown",
          branch: "unknown",
          author_date: now,
          commit_date: now
        }
    end
  end
end

# Always run when script is executed
BuildInfoGenerator.generate()

File: scripts/generate-build-info.js
#!/usr/bin/env node
/**
 * Generic BuildInfo.json generator for React UI apps
 * 
 * Works from any app directory in the monorepo.
 * Generates build-info.json in the app's src/ directory.
 * 
 * Captures:
 * - First 8 digits of latest commit hash
 * - Branch name
 * - Authored date (ISO string, UTC)
 * - Committed date (ISO string, UTC)
 * - Built date (ISO string, UTC)
 * - Mnemonic name (3-word bip39 phrase from commit hash)
 * 
 * Usage:
 *   node scripts/generate-build-info.js [appDir]
 *   If appDir is not provided, detects from current working directory
 */

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

// Determine app directory
const appDir = process.argv[2] || process.cwd();
const appDirPath = path.resolve(appDir);

// Determine repo root (traverse up until .git found)
function findRepoRoot(startPath) {
  let current = path.resolve(startPath);
  while (current !== path.dirname(current)) {
    if (fs.existsSync(path.join(current, '.git'))) {
      return current;
    }
    current = path.dirname(current);
  }
  throw new Error('Could not find repository root (no .git directory found)');
}

const REPO_ROOT = findRepoRoot(appDirPath);

function reverseString(str) {
  return str.split('').reverse().join('');
}

function getBip39WordList() {
  try {
    const bip39Path = path.join(REPO_ROOT, 'libs/utils/src/bip39.ts');
    if (!fs.existsSync(bip39Path)) {
      return null;
    }
    const content = fs.readFileSync(bip39Path, 'utf-8');
    // Extract word list - match the array definition
    const match = content.match(/const bip39WordList = \[([\s\S]*?)\];/);
    if (!match) {
      return null;
    }
    // Parse the array content - split by comma and clean up
    const words = match[1]
      .split(',')
      .map(w => w.trim().replace(/^["']|["']$/g, ''))
      .filter(w => w.length > 0);
    return words;
  } catch (err) {
    console.warn('Warning: Could not load bip39 word list:', err.message);
    return null;
  }
}

function hashToMnemonic(hash, wordList, wordCount = 3) {
  if (!wordList || wordList.length === 0) {
    return null;
  }
  try {
    const reversedHash = reverseString(hash);
    const number = BigInt(`0x${reversedHash}`);
    const mnemonicWords = [];
    let tempNumber = number;
    let count = 0;
    
    while (tempNumber > 0 && count < wordCount) {
      const wordIndex = Number(tempNumber & BigInt(0x7FF)); // 11 bits = 0-2047
      if (wordIndex < wordList.length) {
        mnemonicWords.push(wordList[wordIndex]);
        count++;
      }
      tempNumber = tempNumber >> BigInt(11);
    }
    
    return mnemonicWords.length > 0 ? mnemonicWords.join('-') : null;
  } catch (err) {
    console.warn('Warning: Could not generate mnemonic:', err.message);
    return null;
  }
}

function getGitInfo() {
  try {
    // Always run git commands from repo root to ensure consistency
    const commitHash = execSync('git rev-parse HEAD', { cwd: REPO_ROOT, encoding: 'utf-8' }).trim();
    const shortHash = commitHash.substring(0, 8);
    
    const branch = execSync('git rev-parse --abbrev-ref HEAD', { cwd: REPO_ROOT, encoding: 'utf-8' }).trim();
    
    // Get dates as ISO strings (UTC)
    const authorDate = execSync('git log -1 --format=%aI', { cwd: REPO_ROOT, encoding: 'utf-8' }).trim();
    const commitDate = execSync('git log -1 --format=%cI', { cwd: REPO_ROOT, encoding: 'utf-8' }).trim();
    
    // Generate 3-word mnemonic from commit hash
    const wordList = getBip39WordList();
    const mnemonic = hashToMnemonic(shortHash, wordList, 3) || 'unknown-unknown-unknown';
    
    return {
      commitHash: shortHash,
      branch,
      authoredDate: authorDate,
      committedDate: commitDate,
      mnemonic
    };
  } catch (error) {
    console.warn('Warning: Could not get git info:', error.message);
    const now = new Date().toISOString();
    return {
      commitHash: 'unknown',
      branch: 'unknown',
      authoredDate: now,
      committedDate: now,
      mnemonic: 'unknown-unknown-unknown'
    };
  }
}

function generateBuildInfo() {
  const gitInfo = getGitInfo();
  const builtDate = new Date().toISOString();
  
  const buildInfo = {
    commitHash: gitInfo.commitHash,
    branch: gitInfo.branch,
    authoredDate: gitInfo.authoredDate,
    committedDate: gitInfo.committedDate,
    builtDate: builtDate,
    mnemonic: gitInfo.mnemonic
  };
  
  // Ensure src directory exists
  const srcDir = path.join(appDirPath, 'src');
  if (!fs.existsSync(srcDir)) {
    throw new Error(`Source directory not found: ${srcDir}`);
  }
  
  // Write to src directory so it can be imported
  const outputPath = path.join(srcDir, 'build-info.json');
  fs.writeFileSync(outputPath, JSON.stringify(buildInfo, null, 2));
  
  console.log('BuildInfo.json generated:', outputPath);
  return buildInfo;
}

if (require.main === module) {
  try {
    generateBuildInfo();
  } catch (error) {
    console.error('Error generating build info:', error.message);
    process.exit(1);
  }
}

module.exports = { generateBuildInfo };

